File: assignments/Assignment-04/Multi-Programming Comparison/SRRS_Student_Handover_Package/cpp/src/main.cpp

#include <algorithm>
#include <iostream>
#include <mutex>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

struct Reservation {
    std::string id, user_id, train_id, seat_no;
};

class BookingService {
    std::unordered_map<std::string, std::pair<std::string,std::string>> users{
        {"alice", {"alice123","U100"}}, {"bob", {"bob123","U200"}}
    };
    std::unordered_map<std::string,std::string> trains{{"T1","Aarhus Express"}};
    std::unordered_map<std::string,std::unordered_set<std::string>> seats{
        {"T1", {"S1","S2","S3","S4"}}
    };
    std::unordered_map<std::string,std::string> sessions;
    std::unordered_map<std::string,Reservation> reservations;
    std::mutex seat_mutex;
    int counter = 0;

    std::string next_id(const std::string& prefix) {
        return prefix + std::to_string(++counter);
    }

public:
    std::pair<std::string,std::string> authenticate(
        const std::string& username, const std::string& password) {
        auto it = users.find(username);
        if (it == users.end() || it->second.first != password)
            throw std::runtime_error("authentication_failed");
        std::string token = next_id("TOKEN-");
        sessions[token] = it->second.second;
        return {token, it->second.second};
    }

    std::vector<std::string> search_train(const std::string& criteria) {
        std::vector<std::string> result;

        char search_key[32];
        std::copy(criteria.begin(), criteria.end(), search_key);
        search_key[criteria.size()] = '\0';

        std::string criterion(search_key);
        for (const auto& [id,name] : trains)
            if (id.find(criterion) != std::string::npos ||
                name.find(criterion) != std::string::npos)
                result.push_back(id);
        std::sort(result.begin(), result.end());
        return result;
    }

    std::vector<std::string> available_seats(const std::string& train_id) {
        std::vector<std::string> result;
        auto it = seats.find(train_id);
        if (it == seats.end()) return result;
        for (const auto& s : it->second) result.push_back(s);
        std::sort(result.begin(), result.end());
        return result;
    }

    std::pair<std::string,Reservation> book_seat(
        const std::string& token, const std::string& requested_user,
        const std::string& train_id, const std::string& seat_no) {
        if (sessions.find(token) == sessions.end())
            return {"not_authenticated", {}};

        std::lock_guard<std::mutex> guard(seat_mutex);
        auto it = seats.find(train_id);
        if (it == seats.end() || it->second.erase(seat_no) == 0)
            return {"seat_unavailable", {}};

        Reservation r{next_id("RES-"), requested_user, train_id, seat_no};
        reservations[r.id] = r;
        return {"OK", r};
    }

    std::pair<std::string,Reservation> get_reservation(
        const std::string& token, const std::string& reservation_id) {
        if (sessions.find(token) == sessions.end())
            return {"not_authenticated", {}};
        auto it = reservations.find(reservation_id);
        if (it == reservations.end())
            return {"reservation_not_found", {}};
        return {"OK", it->second};
    }
};

std::string handle(BookingService& service, const std::string& line) {
    std::vector<std::string> p;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, '|')) p.push_back(item);

    try {
        if (p[0] == "AUTH" && p.size() == 3) {
            auto a = service.authenticate(p[1], p[2]);
            return "OK|TOKEN|" + a.first + "|USER|" + a.second;
        }
        if (p[0] == "SEARCH" && p.size() == 2) {
            auto v = service.search_train(p[1]);
            std::string out = "OK|TRAINS|";
            for (size_t i=0;i<v.size();++i) out += (i?",":"") + v[i];
            return out;
        }
        if (p[0] == "SEATS" && p.size() == 2) {
            auto v = service.available_seats(p[1]);
            std::string out = "OK|SEATS|";
            for (size_t i=0;i<v.size();++i) out += (i?",":"") + v[i];
            return out;
        }
        if (p[0] == "BOOK" && p.size() == 5) {
            auto b = service.book_seat(p[1],p[2],p[3],p[4]);
            if (b.first != "OK") return "ERROR|" + b.first;
            auto r = b.second;
            return "OK|RESERVATION|" + r.id + "|USER|" + r.user_id +
                   "|TRAIN|" + r.train_id + "|SEAT|" + r.seat_no;
        }
        if (p[0] == "GET" && p.size() == 3) {
            auto g = service.get_reservation(p[1],p[2]);
            if (g.first != "OK") return "ERROR|" + g.first;
            auto r = g.second;
            return "OK|RESERVATION|" + r.id + "|USER|" + r.user_id +
                   "|TRAIN|" + r.train_id + "|SEAT|" + r.seat_no;
        }
        return "ERROR|invalid_request";
    } catch (...) {
        return "ERROR|request_failed";
    }
}

int main() {
    BookingService service;
    std::string line;
    while (std::getline(std::cin, line)) {
        if (line == "QUIT") break;
        std::cout << handle(service, line) << std::endl;
    }
}

File: assignments/Assignment-04/Multi-Programming Comparison/SRRS_Student_Handover_Package/rust/src/main.rs

use std::collections::{HashMap, HashSet};
use std::io::{self, BufRead};
use std::sync::{Arc, Mutex};

#[derive(Clone)]
struct Reservation {
    id: String,
    user_id: String,
    train_id: String,
    seat_no: String,
}

struct BookingService {
    users: HashMap<String, (String,String)>,
    trains: HashMap<String,String>,
    seats: Arc<Mutex<HashMap<String,HashSet<String>>>>,
    sessions: HashMap<String,String>,
    reservations: HashMap<String,Reservation>,
    counter: u64,
}

impl BookingService {
    fn new() -> Self {
        let mut users = HashMap::new();
        users.insert("alice".into(), ("alice123".into(),"U100".into()));
        users.insert("bob".into(), ("bob123".into(),"U200".into()));

        let mut trains = HashMap::new();
        trains.insert("T1".into(), "Aarhus Express".into());

        let mut all = HashSet::new();
        for s in ["S1","S2","S3","S4"] { all.insert(s.to_string()); }
        let mut seats = HashMap::new();
        seats.insert("T1".into(), all);

        Self {
            users, trains, seats: Arc::new(Mutex::new(seats)),
            sessions: HashMap::new(), reservations: HashMap::new(), counter: 0
        }
    }

    fn id(&mut self, prefix: &str) -> String {
        self.counter += 1;
        format!("{}{}", prefix, self.counter)
    }

    fn authenticate(&mut self, username: &str, password: &str)
        -> Result<(String,String), String> {
        match self.users.get(username) {
            Some((pw, uid)) if pw == password => {
                let uid2 = uid.clone();
                let token = self.id("TOKEN-");
                self.sessions.insert(token.clone(), uid2.clone());
                Ok((token, uid2))
            }
            _ => Err("authentication_failed".into())
        }
    }

    fn search_train(&self, criteria: &str) -> Vec<String> {
        let mut out = Vec::new();
        for (id, name) in &self.trains {
            if id.contains(criteria) || name.contains(criteria) {
                out.push(id.clone());
            }
        }
        out.sort();
        out
    }

    fn available_seats(&self, train_id: &str) -> Vec<String> {
        let seats = self.seats.lock().unwrap();
        let mut out = seats.get(train_id)
            .map(|s| s.iter().cloned().collect::<Vec<_>>())
            .unwrap_or_default();
        out.sort();
        out
    }

    fn book_seat(&mut self, token: &str, requested_user: &str,
                 train_id: &str, seat_no: &str) -> Result<Reservation,String> {
        if !self.sessions.contains_key(token) {
            return Err("not_authenticated".into());
        }

        // Generate the reservation ID before locking the seat store so the
        // mutable operation on self does not overlap the MutexGuard borrow.
        let id = self.id("RES-");

        let mut seats = self.seats.lock().unwrap();
        let set = match seats.get_mut(train_id) {
            Some(s) => s,
            None => return Err("seat_unavailable".into())
        };

        if !set.remove(seat_no) {
            return Err("seat_unavailable".into());
        }
        let r = Reservation {
            id: id.clone(), user_id: requested_user.to_string(),
            train_id: train_id.to_string(), seat_no: seat_no.to_string()
        };
        self.reservations.insert(id, r.clone());
        Ok(r)
    }

    fn get_reservation(&self, token: &str, reservation_id: &str)
        -> Result<Reservation,String> {
        if !self.sessions.contains_key(token) {
            return Err("not_authenticated".into());
        }
        self.reservations.get(reservation_id)
            .cloned()
            .ok_or_else(|| "reservation_not_found".into())
    }
}

fn handle(service: &mut BookingService, line: &str) -> String {
    let p: Vec<&str> = line.trim().split('|').collect();
    if p.is_empty() { return "ERROR|invalid_request".into(); }

    match p[0] {
        "AUTH" if p.len() == 3 => match service.authenticate(p[1],p[2]) {
            Ok((token,user)) => format!("OK|TOKEN|{}|USER|{}",token,user),
            Err(_) => "ERROR|authentication_failed".into()
        },
        "SEARCH" if p.len() == 2 => {
            format!("OK|TRAINS|{}", service.search_train(p[1]).join(","))
        },
        "SEATS" if p.len() == 2 => {
            format!("OK|SEATS|{}", service.available_seats(p[1]).join(","))
        },
        "BOOK" if p.len() == 5 => match service.book_seat(p[1],p[2],p[3],p[4]) {
            Ok(r) => format!("OK|RESERVATION|{}|USER|{}|TRAIN|{}|SEAT|{}",
                             r.id,r.user_id,r.train_id,r.seat_no),
            Err(e) => format!("ERROR|{}",e)
        },
        "GET" if p.len() == 3 => match service.get_reservation(p[1],p[2]) {
            Ok(r) => format!("OK|RESERVATION|{}|USER|{}|TRAIN|{}|SEAT|{}",
                             r.id,r.user_id,r.train_id,r.seat_no),
            Err(e) => format!("ERROR|{}",e)
        },
        _ => "ERROR|invalid_request".into()
    }
}

fn main() {
    let stdin = io::stdin();
    let mut service = BookingService::new();
    for line in stdin.lock().lines() {
        let line = line.unwrap();
        if line.trim() == "QUIT" { break; }
        println!("{}", handle(&mut service, &line));
    }
}
